kratos/containers/constraint_data.h

```
@@ -0,0 +1,352 @@
+        {
+            if ((rObj1->SlaveDofId() == rObj2->SlaveDofId()) && (rObj1->SlaveDofKey() == rObj2->SlaveDofKey()))
+                return true;
+            return false;
+        }
+    };
+
+    typedef std::unordered_set<ConstraintEquationPointerType, EquationHasher, EquationComparator> ConstraintEquationVectorType;
+
+    typedef ConstraintEquationVectorType::iterator iterator;
+
+    ///@name Life Cycle
+    ///@{
+
+    /**
+		Creates a MPC data object
+		*/
+    ConstraintEquationContainer() : mConstraintEquationsVector()
+    {
+    }
+    /// Destructor.
+    virtual ~ConstraintEquationContainer(){};
+
+    ///@}
+
+    ///@name Access
+    ///@{
+
+    /**
+		Clears the maps contents
+		*/
+    void Clear()
+    {
+        mConstraintEquationsVector.clear();
+    }
+    iterator begin() { return mConstraintEquationsVector.begin(); }
+    iterator end() { return mConstraintEquationsVector.end(); }
+
+    /**
+		Get the Constraint equation for this slave
+		@return Data vector for this slave
+		*/
+    ConstraintEquation &GetConstraintEquation(DofType &rSlaveDof)
+    {
+        ConstraintEquationPointerType equation = Kratos::make_shared<ConstraintEquation>(rSlaveDof);
+        auto res = mConstraintEquationsVector.find(equation);
+        if (res != mConstraintEquationsVector.end())
+        {
+            return *(*res);
+        }
+        else
+        {
+            KRATOS_THROW_ERROR("Runtime error :: ","No constraint equation found for given slave dof.\nThere is something worng","");
+        }
+    }
+
+    /**
+		Get the Total number of MasterDOFs for a given slave dof
+		@return Total number of MasterDOFs for a given slave dof
+		 */
+    int GetNumbeOfMasterDofsForSlave(DofType const &rSlaveDof)
+    {
+        ConstraintEquationPointerType equation = Kratos::make_shared<ConstraintEquation>(rSlaveDof);
+        auto res = mConstraintEquationsVector.find(equation);
+        int numMasters = -1;

```

This function expects a slave right? Would it be better to throw an exception if it is not a slave?

---

The SLAVE flag is set on a node and not on a DOF object so there is no possibility to check if a DOF is a slave or not. Unless we store a modelpart ref in this class and access the node from the dof id .